x/superfluid/keeper/migrate_test.go

```diff
@@ -1,253 +1,364 @@
-// 			if tc.percentOfSharesToMigrate.LT(sdk.OneDec()) {
-// 				// Check if the new lock was created.
-// 				newLock, err := lockupKeeper.GetLockByID(ctx, newLockId)
-// 				suite.Require().NoError(err)
-// 				// The new lock should have the same owner and end time.
-// 				// The new lock should have the difference in coins between the original lock and the coins migrated.
-// 				suite.Require().Equal(sdk.NewCoins(poolShareOut.Sub(coinsToMigrate)).String(), newLock.Coins.String())
-// 				suite.Require().Equal(lock.Owner, newLock.Owner)
-// 				suite.Require().Equal(lock.EndTime.String(), newLock.EndTime.String())
-// 				// If original lock was unlocking, the new lock should also be unlocking.
-// 				if lock.IsUnlocking() {
-// 					suite.Require().True(newLock.IsUnlocking())
-// 				}
-// 			} else {
-// 				suite.Require().Equal(uint64(0), newLockId)
-// 			}
-
-// 			// Additional checks if the lock was superfluid staked.
-// 			if tc.superfluidDelegated {
-// 				// Check if migration deleted intermediary account connection.
-// 				addr := superfluidKeeper.GetLockIdIntermediaryAccountConnection(ctx, lockID)
-// 				suite.Require().Equal(addr.String(), "")
-
-// 				// Check if migration deleted synthetic lockup.
-// 				_, err = lockupKeeper.GetSyntheticLockup(ctx, lockID, keeper.StakingSyntheticDenom(lock.Coins[0].Denom, valAddr.String()))
-// 				suite.Require().Error(err)
-
-// 				// Check if delegation has reduced from intermediary account.
-// 				delegation, found := stakingKeeper.GetDelegation(ctx, intermediaryAcc.GetAccAddress(), valAddr)
-// 				suite.Require().False(found, "expected no delegation, found delegation w/ %d shares", delegation.Shares)
-// 			}
-// 		})
-// 	}
-// }
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	cltypes "github.com/osmosis-labs/osmosis/v15/x/concentrated-liquidity/types"
+	"github.com/osmosis-labs/osmosis/v15/x/gamm/pool-models/balancer"
+	gammtypes "github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+	lockuptypes "github.com/osmosis-labs/osmosis/v15/x/lockup/types"
+	"github.com/osmosis-labs/osmosis/v15/x/superfluid/keeper"
+	"github.com/osmosis-labs/osmosis/v15/x/superfluid/types"
+	superfluidtypes "github.com/osmosis-labs/osmosis/v15/x/superfluid/types"
+)
+
+// We test migrating in the following circumstances:
+// 1. Migrating lock that is not superfluid delegated, not unlocking.
+// 2. Migrating lock that is not superfluid delegated, unlocking.
+// 3. Migrating lock that is superfluid delegated, not unlocking.
+// 4. Migrating lock that is superfluid undelegating, not unlocking.
+// 5. Migrating lock that is superfluid undelegating, unlocking.
+func (suite *KeeperTestSuite) TestMigrateLockedPositionFromBalancerToConcentrated() {
```

Comment:
I agree with Matt that the current testing strategy is not sufficient. 

It's also difficult to reason about the test case that has so many side effects.

Ideally, the test for the wrapper router should be basic, verifying that it routes to the right underlying method. And each method individually should be verified to exhibit the desired behavior. This will make it much easier to reason about each test case and confirm that the testing strategy used is sufficient.

Would like to bump the priority on this: https://github.com/osmosis-labs/osmosis/issues/4977 if we are punting it till later. Also, would appreciate if you could port the context from this comment to #4977

Reply:
Added this context to the issue itself